router/h2/src/main/scala/io/buoyant/router/h2/ClassifierFilter.scala

```diff
@@ -11,40 +14,71 @@ object ClassifierFilter {
-      val role = ClassifierFilter.role
-      val description = "Sets the response classification into a header"
-      def make(classifierP: param.ResponseClassifier, next: ServiceFactory[Request, Response]) = {
-        val param.ResponseClassifier(classifier) = classifierP
+    new Stack.Module1[param.H2Classifier, ServiceFactory[Request, Response]] {
+      override val role: Stack.Role = ClassifierFilter.role
+      override val description = "Sets the stream classification into a header"
+      override def make(
+        classifierP: param.H2Classifier,
+        next: ServiceFactory[Request, Response]
+      ): ServiceFactory[Request, Response] = {
+        val param.H2Classifier(classifier) = classifierP
         new ClassifierFilter(classifier).andThen(next)
       }
     }
 
-  val successClassClassifier: ResponseClassifier = {
-    case ReqRep(req, Return(rep: Response)) if rep.headers.contains(SuccessClassHeader) =>
-      val success = rep.headers.get(SuccessClassHeader)
-        .flatMap { value =>
-          Try(value.toDouble).toOption
-        }.getOrElse(0.0)
-      if (success > 0.0)
-        ResponseClass.Successful(success)
-      else
-        ResponseClass.Failed(false)
+  private[this] object ResponseSuccessClass {
+    @inline def unapply(headers: Headers): Option[ResponseClass] =
+      headers.get(SuccessClassHeader).map { value =>
+        val success = value.toDouble
+        if (success > 0.0) ResponseClass.Successful(success)
+        else ResponseClass.Failed(false)
+      }
   }
-}
 
-class ClassifierFilter(classifier: ResponseClassifier) extends SimpleFilter[Request, Response] {
+  object SuccessClassClassifier extends H2Classifier {
+    override val streamClassifier: PartialFunction[H2ReqRepFrame, ResponseClass] = {
+      case H2ReqRepFrame(_, Return((_, Some(Return(ResponseSuccessClass(c)))))) => c
+    }
 
-  private[this] val successClass = classifier.andThen(_.fractionalSuccess).lift
+    override val responseClassifier: PartialFunction[H2ReqRep, ResponseClass] = {
+      case H2ReqRep(_, Return(rsp)) if rsp.headers.contains(SuccessClassHeader) =>
+        rsp.headers.get(SuccessClassHeader).map { value =>
+          val success = value.toDouble
+          if (success > 0.0) ResponseClass.Successful(success)
+          else ResponseClass.Failed(false)
+        }.get
+    }
+  }
+}
 
+class ClassifierFilter(classifier: H2Classifier) extends SimpleFilter[Request, Response] {
+  import ClassifierFilter.SuccessClassHeader
+  private[this] val successHeader: ResponseClass => String =
+    _.fractionalSuccess.toString
   def apply(req: Request, svc: Service[Request, Response]): Future[Response] = {
-    svc(req).map { rep =>
-      successClass(ReqRep(req, Return(rep))) match {
-        case Some(success) =>
-          rep.headers.set(ClassifierFilter.SuccessClassHeader, success.toString)
-        case None =>
+    svc(req).map { rep: Response =>
+      if (rep.stream.isEmpty) {
```

Comment:
I'm not sure if this will work like that – in order to call `responseClassifier.isDefinedAt`, I'd have to take the `Response` and wrap it in a `H2ReqRep`, and whether or not the response also has a non-empty stream, the `responseClassifier` will be defined...